### DISCUSSION TITLE
Ad hoc evaluation of input given a policy

### DISCUSSION THREAD

**Discussion by firstdorsal** [open-policy-agent/community#72] [OPA and Rego] [answered]
Hi!
I want to give the OPA REST API a policy:

```rego
package play
import future.keywords.in

default domain = false
default apiMethods = false
default rrTypes = false
default valuePatterns = false


domain {
    startswith(input.domain,"_acme-challenge.")
    endswith(input.domain,".")
}

apiMethods {
    input.apiMethods in ["set","get","delete"]
}

rrTypes {
    input.rrTypes in ["TXT"]
}

valuePatterns {
    true
}
```

and a query:

```json
{
        "domain": "_acme-challenge.y.gy.",
        "apiMethods": "set",
        "rrTypes": "TXT",
        "valuePatterns": "any"
}
```

and want it to evaluate the input based on the policy like it can be seen on the playground.
See the working playground example here: https://play.openpolicyagent.org/p/1jbVAVNrtu

I tried doing exactly that with the REST API like that:
```sh
curl http://[::]:8081/v1/query -X "POST" -v -H "Content-Type: application/json" \
--data-binary @- << EOF
{
  "query": "package play\nimport future.keywords.in\n\ndefault domain = false\ndefault apiMethods = false\ndefault rrTypes = false\ndefault valuePatterns = false\n\n\ndomain {\n    startswith(input.domain,\"_acme-challenge.\")\n    endswith(input.domain,\".\")\n}\n\napiMethods {\n    input.apiMethods in [\"set\",\"get\",\"delete\"]\n}\n\nrrTypes {\n    input.rrTypes in [\"TXT\"]\n}\n\nvaluePatterns {\n    true\n}",
  "input": {
    "domain":"_acme-challenge.y.gy.",
    "apiMethods":"set",
    "rrTypes":"TXT",
    "valuePatterns": "any"
  }
}
EOF
```
with the response:
```sh
{
  "code": "internal_error",
  "message": "expected body but got *ast.Package"
}
```

I figured out that it is possibly only evaluating single lines, but this is not pointed out in the documentation.

I looked at playgrounds HTTP request, but I think these are doing something else in the backend than the standard REST API because I couldn't reproduce these either.

Is it possible to do what I want with one request?
Would I have to send multiple requests? (Create policy, send query, delete policy)

Am I missing something?

Thank you!

**Comment by firstdorsal** [comment, marked as answer]
this is my solution for now:

it took forever to find out how it works... the documentation is everything but not helpful :(

via the cli
```sh
opa eval --data acme.rego --input input.json data.system
```

via rest api:
```sh
curl http://[::]:8081/v1/policies/x -X "PUT" -v -H "Content-Type: text/plain" --data-binary @acme.rego

curl http://[::]:8081/ -X "POST" -v -H "Content-Type: application/json" --data-binary @input.json
```

for the files:
**acme.rego**

*Note that naming the package is done in the rego file! Here: `system.main`*
*When then contacting the root url: `http://[::]:8081/` it queries the package `system.main` per default*

```rego
package system.main
import future.keywords.in

default domain = false
default apiMethods = false
default rrTypes = false
default valuePatterns = false


domain {
    startswith(input.domain,"_acme-challenge.")
    endswith(input.domain,".")
}

apiMethods {
    input.apiMethods in ["set","get","delete"]
}

rrTypes {
    input.rrTypes in ["TXT"]
}

valuePatterns {
    true
}
```
and 
**input.json**
```json
{
    "domain": "_acme-challenge.y.gy.",
    "apiMethods": "set",
    "rrTypes": "TXT",
    "valuePatterns": "any"
}
```

**Reply by challarao** [reply]
Hey @firstdorsal, I'm looking out for clearing similar confusion. In your post, you mentioned about root URL querying system.main. 

```
Note that naming the package is done in the rego file! Here: system.main
When then contacting the root url: http://[::]:8081/ it queries the package system.main per default
```
How to query non system.main package(like play package)? Does it accept any params or is the Data API only way to do that for non system packages?

**Reply by firstdorsal** [reply]
I would love to help but I am not using OPA anymore and have no good understanding of it.

**Reply by challarao** [reply]
@anderseknert any thoughts?

**Reply by anderseknert** [reply]
You can query any package and rule from the `system.main` package. See [this blog](https://blog.styra.com/blog/dynamic-policy-composition-for-opa) for some ideas on how you could do that.

**Comment by anderseknert** [comment]
You'll want to query the [data API](https://www.openpolicyagent.org/docs/latest/rest-api/#data-api). Given you original policy:

```rego
package play

import future.keywords.in

# ...
```

You'd then query OPA at the `/v1/data/play` endpoint (matching the package name in this case, though you could also query an individual rule as well, like `/v1/data/play/domain`):

`curl http://localhost:8081/v1/data  -d"@input.json"`

The [HTTP API](https://www.openpolicyagent.org/docs/latest/http-api-authorization/) docs provide a pretty good guide to running and querying OPA in a scenario like this.

**Reply by firstdorsal** [reply]
Thank you for the quick response!

Yes doing it in two request is stated but why isnt it possible to send this policy as adhoc query?
with `package system.main` or any package it throws:
```json
{
  "code": "internal_error",
  "message": "expected body but got *ast.Package"
}
```

and when removing the package it throws
```json
{
  "code": "internal_error",
  "message": "expected body but got *ast.Import"
}
```
It seems to only evaluate single statements and (i think) this  is not sufficiently stated in the documentation. 
I would expect an ad hoc evaluation to take my policy and my input to return an answer.

**Reply by anderseknert** [reply]
Ah, OK! I wasn't sure the single request was what you meant by ad hoc, but that makes sense. No, you currently can't upload a policy and have it evaluated within a single request. If you'd want to develop something like the Rego Playground you'd currently need to use the OPA Go API for that.

**Reply by firstdorsal** [reply]
It would be awesome to have that as a feature at some point :D
Thank you very much and keep up the great work!

**Reply by firstdorsal** [reply]
Would it be possible to get an open source version of the rego playground?

**Reply by anderseknert** [reply]
I'm not aware of any plans to open source the Rego playground, I'm afraid.
